### PR TITLE
app: runCommand: Add aws to the allowlist for EKS/SSO workflows

### LINKED ISSUE #5662: runCommand: add 'aws' to the allowlist for EKS/SSO workflows
## Summary

The `runCommand` API currently allows only `minikube`, `az`, and `scriptjs`. This prevents plugins from triggering `aws sso login` when an AWS SSO session expires, which is a very common workflow for anyone using Headlamp with Amazon EKS.

## Use case

AWS EKS is one of the most widely used Kubernetes platforms. Users who authenticate via AWS SSO (IAM Identity Center) need to periodically re-run `aws sso login` when their session expires. Without `aws` in the allowlist, a Headlamp plugin can detect the expired session (via a 401/403 response) but cannot do anything about it beyond displaying a message and asking the user to copy-paste the command into a terminal.

A plugin that could call `runCommand('aws', ['sso', 'login'])` on the user's behalf would close this loop entirely — detect expiry → prompt user → re-authenticate → continue — without leaving Headlamp.

## Current workaround

Plugins can show a chip/dialog with a copy button for `aws sso login`. This works but adds unnecessary friction, especially for less technical users.

## Request

Add `aws` to the `runCommand` allowlist (alongside `minikube` and `az`), so that plugins targeting AWS EKS clusters can trigger re-authentication flows natively.

If there are security concerns about arbitrary `aws` subcommands, a scoped allowlist (e.g. only `aws sso login`) would also be acceptable.

## References

- [runCommand docs / registry.ts](https://github.com/headlamp-k8s/headlamp/blob/main/src/plugin/registry.tsx)
- AWS EKS is the [most widely adopted managed Kubernetes service](https://www.cncf.io/reports/cncf-annual-survey-2023/)

### PR DESCRIPTION
**Pull request by bhuvan-somisetty**
## Summary

This PR adds `aws` to the `runCommand` allowlist so the official AI Assistant plugin can trigger `aws sso login` to help users re-authenticate when their AWS SSO session expires on EKS clusters.

## Related Issue

Fixes #5662

## Context on the maintainer's question

In the issue, @illume asked:

> We limit these to single official plugins as well as an extra required security measure. Do you have a particular plugin in mind?

This PR keeps that security model intact:

- `aws` is added to the base command allowlist (`validateCommandData` in `app/electron/runCmd.ts`), same as `az`/`gh`/`minikube`.
- `pluginRunCommand` access to `aws` is only granted to the official `@headlamp-k8s/ai-assistant` plugin (`frontend/src/plugin/index.ts`), the same plugin that already has `az`/`gh` access — no other plugin gets this permission.
- Only `aws sso login` is added to that plugin's pre-consented command list (`COMMANDS_WITH_CONSENT.headlamp_ai_assistant`). Any other `aws` subcommand still requires the existing interactive consent dialog before it can run — nothing runs without the user's consent.

This matches the scoped-allowlist compromise suggested in the issue itself ("a scoped allowlist (e.g. only `aws sso login`) would also be acceptable").

## Changes

- `app/electron/runCmd.ts`: Added `aws` to `validCommands` in `validateCommandData`; added `aws sso login` to `COMMANDS_WITH_CONSENT.headlamp_ai_assistant`
- `frontend/src/components/App/runCommand.ts`: Added `aws` to the `runCommand` command type union
- `frontend/src/plugin/index.ts`: Granted `runCmd-aws` permission secret and `aws` command access to the `@headlamp-k8s/ai-assistant` plugin only
- `app/electron/runCmd.test.ts`: Added a test verifying `aws` is accepted by `validateCommandData`
- `app/electron/runCmd.ts`: Fixed `checkCommandConsent` returning `true` even when the user denies a command at the prompt — it saved the denial but still let that first run through. This is pre-existing on main, but every `aws` subcommand outside the pre-consented `aws sso login` relies on that dialog to gate it, so the scoping described above only holds once it is fixed.

## Steps to Test

1. Run the existing test suites: `cd app && npx vitest run electron/runCmd.test.ts` and `cd frontend && npx vitest run src/plugin/runPlugin.test.ts`
2. Confirm `validateCommandData` accepts `{ command: 'aws', args: ['sso', 'login'], ... }` and still rejects any command not in the allowlist
3. Confirm no plugin other than `@headlamp-k8s/ai-assistant` receives the `runCmd-aws` permission secret or the `aws` option in `pluginRunCommand`

## Screenshots (if applicable)

N/A — backend/plugin-permission change, no UI change.

## Notes for the Reviewer

- No new plugin is introduced by this PR; it only extends the existing `@headlamp-k8s/ai-assistant` plugin's permission scope, matching how it already has `az`/`gh` access for similar auth-refresh workflows.
- Any `aws` subcommand outside `aws sso login` still triggers the existing per-command consent dialog (`confirmCommandDialog`) the first time it's used — this PR does not bypass user consent for arbitrary `aws` usage.

## Summary by CodeRabbit

* **New Features**
  * Added support for running AWS commands.
  * Added consent handling for AWS commands, including AWS SSO login.
  * Improved command permission prompts and saved consent behavior.

* **Bug Fixes**
  * Commands are now blocked immediately when consent is denied.
  * Improved validation for command IDs, arguments, options, permission data, and saved settings.
  * Preserved explicit permission denials when updating command consent.

### CODE REVIEW

**Comment by bhuvan-somisetty** [comment]
@illume this addresses the security-scoping concern you raised on the issue — aws access is scoped only to the official ai-assistant plugin, and only the aws sso login subcommand is pre-consented, matching the az/gh pattern already in place. All CI checks are green. Would appreciate a review when you have a chance, thanks!

**Comment by ashu8912** [comment]
Hii @bhuvan-somisetty what's the reason for this PR? What usecase does it suit for ai-assistant plugin? Also any particular flow you tested this on??

**Comment by bhuvan-somisetty** [comment]
Hi @ashu8912, thanks for taking a look!

**Reason for this PR / use case:** The `@headlamp-k8s/ai-assistant` plugin already has scoped `runCommand` access to `az` and `gh` so it can help users re-authenticate to Azure/GitHub when their session expires mid-workflow. This PR extends the same pattern to `aws`, specifically to let the assistant trigger `aws sso login` so users on EKS clusters can refresh an expired AWS SSO session without leaving Headlamp. It was originally scoped down in response to @illume's security concern on the linked issue (#5662) to only the `aws sso login` subcommand, for the official ai-assistant plugin only.

**What I tested:** Re-running the unit suites (`app/electron/runCmd.test.ts`, `frontend/src/plugin/runPlugin.test.ts`) and manually exercising `validateCommandData`/`checkPermissionSecret` for the `aws` command. While re-checking this after the branch update, I actually found (and just fixed in the latest commit) two real bugs that would have broken this flow in practice:

1. The Electron main process never generated a `runCmd-aws` permission secret in `setupRunCmdHandlers`, so any `aws` command from the renderer would always be rejected by `checkPermissionSecret`.
2. `checkCommandConsent` only used the first argument to build its consent key (e.g. `"aws sso"`), while the pre-consent list stored `"aws sso login"`. The keys never matched, so the intended pre-consent silently would not have applied, and if it had matched loosely it would've covered every `aws sso` subcommand instead of just `login` as described in the PR.

Both are fixed now: the main process generates a `runCmd-aws` secret, and the consent-key builder includes the second arg specifically for `aws` so it matches `"aws sso login"` exactly (verified with new tests). All checks are green again after rebasing on the updated branch.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for this PR.


the PR has a merge-main commit; please rebase against main to keep the history clean.

<details>
<summary>Why this matters</summary>

Merge commits from `main` make the PR history harder to review. Please rebase your branch on top of the latest `main` instead, then update the PR with the rebased commits.

</details>


A few of the commits don't quite follow the project guidelines. We use Linux kernel style for git commits — have a look at the [contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) and previous commits with `git log`.

<details>
<summary>Commits that need attention</summary>

- `app: runCommand: fix aws sso login permission secret and consent scoping` — Description must start with a capital letter — e.g. `frontend: HomeButton: Fix the button` not `frontend: HomeButton: fix the button`.

</details>

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <Description of changes>` — description must start with a capital letter.
- Keep the title under 72 characters (soft requirement).
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

**Comment by bhuvan-somisetty** [comment]
@illume Thanks for the review! Rebased on top of the latest main to drop the merge commit, and fixed the commit message so the description starts with a capital letter. History should be clean now. Let me know if anything else needs a look.

**Comment by vtushar06** [comment]
traced the consent path here and it holds end to end: addRunCmdConsent seeds `aws sso login` and checkCommandConsent builds the same three-part key via the aws special-case, so they match and it won't re-prompt. the runCmd-aws secret is wired on the electron side too, and there are tests for both the match and the distinct-key case for other aws sso subcommands. consent and secret side looks solid to me @illume.

**Comment by bhuvan-somisetty** [comment]
Thanks for verifying that, @vtushar06! Appreciate you tracing through the key-matching logic.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


Would you mind addressing the open Copilot review comments? Please mark each comment as resolved after addressing it.

**Comment by bhuvan-somisetty** [comment]
Thanks @illume — addressed all 4 open Copilot comments and marked them resolved:

- Consent key mismatch and missing `runCmd-aws` secret were already fixed in the last push.
- Found one more real issue while re-checking: `addRunCmdConsent` used a truthiness check (`if (!settings.confirmedCommands[command])`) when seeding pre-consent entries on plugin install, so an explicit prior denial (`false`) would get silently overwritten back to `true` on reinstall. Changed it to check for `undefined` instead so an existing user choice is never overridden.
- Added a test covering `addRunCmdConsent` seeding `aws sso login` on install and preserving an existing denial.

All tests/typecheck/lint pass locally, squashed into the existing fix commit to keep history clean. CI running now.

**Comment by ashu8912** [comment]
Hii @bhuvan-somisetty i don't think the az cli and gh cli serve that purpose here, @illume would you like to add comments to this

**Comment by bhuvan-somisetty** [comment]
Thanks! i'll wait for @illume's feedback and make any requested changes

**Comment by bhuvan-somisetty** [comment]
Fair point @ashu8912 — agreed the `az`/`gh` comparison isn't exact. The real justification is the use case in #5662, independent of that analogy: users on EKS who authenticate via AWS SSO have to re-run `aws sso login` when their session expires, and today a plugin can detect the expired session (401/403) but can only tell the user to run the command in a terminal. Letting the assistant call `runCommand('aws', ['sso', 'login'])` closes that loop in-app.

It's scoped narrowly — `aws` access is granted only to the official `@headlamp-k8s/ai-assistant` plugin, and only `aws sso login` is pre-consented; any other `aws` subcommand still hits the interactive consent dialog. Happy to hear your and @illume's thoughts on the direction.

**Comment by bhuvan-somisetty** [comment]
@illume @ashu8912 this is ready for another review whenever you have a chance — all CI checks are green, the open Copilot comments have been resolved, and I've replied to the scoping question above. Happy to make any further changes you'd like. Thanks!

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for this PR.


the PR has a merge-main commit; please rebase against main to keep the history clean.

<details>
<summary>Why this matters</summary>

Merge commits from `main` make the PR history harder to review. Please rebase your branch on top of the latest `main` instead, then update the PR with the rebased commits.

</details>

**Comment by bhuvan-somisetty** [comment]
@illume thanks for the review, and sorry about the merge commits.

Rebased onto latest `main` and force-pushed — the branch is now just the two original commits (`bcbe06a2`, `7bf901e3`) with no merge commits. The diff is unchanged from before the rebase, and CI is green.

Ready for another look whenever you have time.

File: app/electron/runCmd.ts [hunk 1]
```diff
@@ -75,14 +75,24 @@ function confirmCommandDialog(command: string, mainWindow: BrowserWindow): boole
...
+  // "aws sso" call, keeping the allowlist as narrow as intended.
   let consentKey = command;
   if (args && args.length > 0) {
     consentKey += ' ' + args[0];
+    if (command === 'aws' && args.length > 1) {
+      consentKey += ' ' + args[1];
+    }
   }
 
   const savedCommand: boolean | undefined = confirmedCommands
```
[the remark this answers is not in the document]
Comment (bhuvan-somisetty) [reply app/electron/runCmd.ts:116]: Good catch, this one is real. The dialog result was saved but the function fell through to `return true`, so a denial only took effect from the *next* call onwards — the command still ran once.

Fixed in 972119dc by returning the choice the user just made. Added tests for both the deny and allow paths at the prompt; I confirmed the deny test fails against the old code and passes now.

Worth noting for reviewers that this is pre-existing on main rather than something this PR introduced. I fixed it here anyway because every `aws` subcommand other than the pre-consented `aws sso login` depends on this dialog to gate it, so leaving it would undercut the scoping the rest of the PR is built on.

**Comment by bhuvan-somisetty** [comment]
@iilume could you take a look at this when you get a chance? Would appreciate your review.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


Would you mind addressing the open Copilot review comments? Please mark each comment as resolved after addressing it.

File: app/electron/runCmd.ts [hunk 2]
```diff
@@ -428,7 +442,7 @@ export function validateCommandData(eventData: CommandDataPartial): [boolean, st
     }
   }
 
-  const validCommands = ['minikube', 'az', 'scriptjs', 'gh'];
+  const validCommands = ['minikube', 'az', 'aws', 'scriptjs', 'gh'];
```
[the remark this answers is not in the document]
Comment (bhuvan-somisetty) [reply app/electron/runCmd.ts:510]: Good catch. Hardened validateCommandData to reject args arrays that contain non-string elements and to reject options/permissionSecrets when they are arrays (typeof [] === 'object' was letting those through). Added tests for both cases.

**Comment by bhuvan-somisetty** [comment]
Hi @illume, I have squashed the commits, addressed the feedback on validation, and updated the branch against the latest main. All checks are passing. Whenever you have time, please review. Thank you!

**Comment by bhuvan-somisetty** [comment]
Hi @illume, all Copilot review comments and threads have been addressed and resolved. Copilot generated no new comments. All checks and unit tests are passing.

**Comment by bhuvan-somisetty** [comment]
Hi @illume, I have updated the consent handlers (\checkCommandConsent\, \ddRunCmdConsent\, \
emoveRunCmdConsent\) to defensively normalize \settings.confirmedCommands\ so corrupted/non-object settings gracefully prompt for consent without throwing. Added a unit test covering corrupted settings in \
unCmd.test.ts\. All checks and tests are passing.

**Comment by bhuvan-somisetty** [comment]
Rebased on top of the latest `main` to resolve the merge conflict in `app/electron/runCmd.test.ts` (merging in the shell-environment mocks/tests from main alongside this PR's aws-allowlist test additions). Also resolved the one outstanding Copilot thread about normalizing a corrupted `confirmedCommands` — that guard is already present at every write site (`checkCommandConsent`, `addRunCmdConsent`, `removeRunCmdConsent`). All 37 unit tests in `runCmd.test.ts` and the full type-check pass locally. Should be mergeable now.

**Comment by bhuvan-somisetty** [comment]
@illume this one's rebased and passing CI too, would appreciate your eyes on it whenever you're free

**Comment by bhuvan-somisetty** [comment]
Addressed Copilot's suppressed comments: the denial-path log said 'Invalid command' when the command was actually valid and just explicitly denied by the user, so I reworded it to 'Command denied by user' to avoid confusing that with an allowlist failure during troubleshooting. Also fixed the two test reset blocks that were using gh/az as consent keys instead of the real 'gh auth'/'az account' format the code actually generates, so they now match production behavior. All 37 tests in runCmd.test.ts pass.

**Comment by bhuvan-somisetty** [comment]
Copilot flagged one more thing: validateCommandData wasn't checking eventData.id at all, even though handleRunCommand uses it to tag every command-stdout/stderr/exit IPC message. A missing or non-string id would've broken output routing on the renderer side. Added the check plus a test for it, and updated the existing 'returns true' test cases to include an id since they'd now fail validation without one. All 38 tests pass.

**Comment by bhuvan-somisetty** [comment]
Fixed — those three negative tests were missing \`id\`, so they were tripping the new required-id check before ever exercising the assertion they were meant to test (args, options, permissionSecrets as arrays). Added \`id: 'test-id'\` to each so they actually validate what they claim. All tests pass.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for this PR.


There are git conflicts in this PR that need to be resolved.

<details>
<summary>How to resolve conflicts</summary>

Rebase or merge the latest main into your branch, resolve the conflicts, and push the updated branch.

</details>

**Comment by bhuvan-somisetty** [comment]
Hi @illume, I've rebased the branch onto latest main and resolved the conflict in app/electron/runCmd.test.ts. All unit tests and checks are passing cleanly.

**Comment by bhuvan-somisetty** [comment]
@illume rebased onto main and resolved the conflict, and cleaned up the last round of Copilot feedback: settings.json can technically parse to null/an array/etc if it's corrupted, and checkCommandConsent/addRunCmdConsent/removeRunCmdConsent were dereferencing `settings.confirmedCommands` without checking `settings` itself first, which would throw. Pulled that normalization into a shared `asSettingsObject` helper (also used it to dedupe the copy that already existed in tray.ts). Also fixed a couple of test issues: some validateCommandData negative tests were missing an `id`, so they were short-circuiting on the id check instead of testing what they claimed to; and the "corrupted value" consent test was inheriting call counts from the previous test's dialog mock. Everything's green — tests, tsc, and the merge with main.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


would you mind rebasing this branch on top of the latest main rather than merging it in?

<details>
<summary>Why this matters</summary>

Merge commits from `main` make the PR history harder to review. Please rebase your branch on top of the latest `main` instead, then update the PR with the rebased commits.

</details>


The commit messages could use some tidying up to match our contribution guidelines. We use Linux kernel style — the [contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) has the details, and `git log` shows good examples.

<details>
<summary>Commits that need attention</summary>

- `runCmd: address latest Copilot review feedback on PR 6449` — Description must start with a capital letter — e.g. `frontend: HomeButton: Fix the button` not `frontend: HomeButton: fix the button`.

</details>

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <Description of changes>` — description must start with a capital letter.
- Keep the title under 72 characters (soft requirement).
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for these changes.


can you please rebase against main to remove the merge main commit?

<details>
<summary>Why this matters</summary>

Merge commits from `main` make the PR history harder to review. Please rebase your branch on top of the latest `main` instead, then update the PR with the rebased commits.

</details>


Can you please have a look at the git commits to see if they meet the contribution guidelines? We use a Linux kernel style of git commits. See the [contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) for general context, and please see previous git commits with git log for examples.

<details>
<summary>Commits that need attention</summary>

- `runCmd: address latest Copilot review feedback on PR 6449` — Description must start with a capital letter — e.g. `frontend: HomeButton: Fix the button` not `frontend: HomeButton: fix the button`.

</details>

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <Description of changes>` — description must start with a capital letter.
- Keep the title under 72 characters (soft requirement).
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

**Comment by bhuvan-somisetty** [comment]
Fixed the real bug Copilot caught: checkCommandConsent was persisting a denial to settings.json whenever there was no window to actually show the dialog on, which would permanently block the command even after a window became available. Now it just denies that one call without writing anything. Also pulled the repeated "is this a plain object" check into a shared isPlainObject helper and fixed the permissionSecrets error message so it doesn't blame typeof for an array. Added a regression test for the null-window case. Tests and tsc are green.

**Comment by bhuvan-somisetty** [comment]
Rebased onto main and squashed the review-correction commits into one — the merge commit and the separate "address feedback" commits were making it hard to review as a coherent change. History's linear now, single commit, force-pushed. Tests and tsc still green.

**Comment by bhuvan-somisetty** [comment]
Good catch, added a regression test for this. Pulled the permission/wrapper selection out of the closure into `getAllowedRunCmdPermissions` in runPlugin.ts so it's directly testable, and added cases covering: the official ai-assistant package getting gh/az/aws, a plugin with the right name but wrong path (or vice versa) getting nothing, an arbitrary plugin getting no runCmd secrets, and minikube's secrets staying scoped separately from ai-assistant's. All green now.

**Comment by bhuvan-somisetty** [comment]
Thanks for the thorough pass. Fixed both:

- The AWS grant really was only checking two attacker-controlled strings (folder path + package.json name), so a look-alike plugin published under a different ArtifactHub repo could impersonate the official ai-assistant/minikube packages. `getAllowedRunCmdPermissions` now also checks provenance — shipped/dev plugins are trusted by filesystem location, and user-installed ones have to match the official `headlamp-plugins` ArtifactHub repo (from the installer-set `artifacthub.repoName`), not just the name they claim. Added tests for the look-alike case specifically.
- Also squashed the two commits back into one so the implementation and its tests land together, and forced piped stdio on the spawned child so a plugin can't crash the main process by passing `stdio: 'ignore'`/`'inherit'` through options.

All tests green, force-pushed.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


Would you mind addressing the open Copilot review comments? Please mark each comment as resolved after addressing it.

**Review by skoeva — COMMENTED** [review]
setting this to draft until we have an update on the use case

**Comment by bhuvan-somisetty** [comment]
Addressed all Copilot review comments and resolved the review conversations . All unit tests passing cleanly.
